asf_public_discourse_home_decarbonisation/utils/plotting_utils.py

```diff
@@ -0,0 +1,466 @@
+
+
+######### 1. Distribution of posts over time #########
+def plot_post_distribution_over_time(
+    dataframe: pd.DataFrame, output_path: str, color: str
+) -> None:
+    """
+    Analyse and plot the distribution of posts over time.
+
+    Parameters:
+        dataframe : pd.DataFrame
+            The DataFrame containing the data.
+        output_path : str
+            The path to save the plot.
+        color : str
+            Color for the plot.
+
+    Returns:
+        None: This function does not return anything. It generates and saves a plot.
+    """
+
+    sample_size = dataframe.shape[0]
+    dataframe["year"] = dataframe["date"].dt.year
+    post_count_by_year = dataframe.groupby("year").size().reset_index(name="post_count")
+
+    min_year = post_count_by_year["year"].min()
+    max_year = post_count_by_year["year"].max()
+    all_years = pd.DataFrame({"year": list(range(min_year, max_year + 1))})
+    post_count_by_year = pd.merge(
+        all_years, post_count_by_year, on="year", how="left"
+    ).fillna(0)
+
+    plt.figure(figsize=(12, 8))
+    sns.barplot(x="year", y="post_count", data=post_count_by_year, color=color)
+    plt.title(
+        f"Distribution of Posts Over Time (By Year) (total sample size n = {sample_size})"
+    )
+    plt.xlabel("Year")
+    plt.ylabel("Number of Posts")
+    plt.tight_layout()
+    plt.savefig(os.path.join(output_path, "Distribution_of_Posts_By_Year.png"))
+    plt.show()
+
+
+########## 2. Plot of the number of users vs number of posts ###########
+
+
+def plot_users_post_distribution(dataframe, output_path):
+    """
+    Creates and saves a plot showing the distribution of the number of posts vs the number of users.
+
+    Args:
+        dataframe (pandas.DataFrame): The DataFrame containing user post data with a 'username' column.
+        output_path (str): The directory path to save the output plot.
+
+    Returns:
+        None: The function saves the plot as a PNG file and displays it.
+    """
+    # Calculate the post counts per user
+    user_post_counts = dataframe["username"].value_counts()
+    sample_size = user_post_counts.size
+    print("sample size:")
+    print(sample_size)
+    # Count how many users have the same number of posts
+    post_count_distribution = user_post_counts.value_counts().sort_index()
+    plt.figure(figsize=(12, 8))
+    # Apply the Freedman-Diaconis Rule to determine bin size
+    iqr = np.subtract(*np.percentile(post_count_distribution, [75, 25]))
+    multiplier = 2.5  # Increase this for larger bins
+    bin_width_fd = multiplier * iqr * (sample_size ** (-1 / 3))
+    bin_size_fd = math.ceil(
+        (max(post_count_distribution) - min(post_count_distribution)) / bin_width_fd
+    )
+    max_bins = 50  # Adjust as needed
+    bin_size_fd = min(bin_size_fd, max_bins)
+    post_range = max(post_count_distribution.index) - min(post_count_distribution.index)
+    number_of_bins = post_range / bin_size_fd
```

Comment:
The Freedman-Diaconis rule is good for defining bin width in histograms because it's data-driven, robust to outliers, and effective in balancing the granularity of the histogram, making it suitable for a wide range of data types and sizes! https://en.wikipedia.org/wiki/Freedman%E2%80%93Diaconis_rule